README.txt
README.TXT -  CSU/CIRA/RAMMB Hurricane Intensity and Strength Algorithm (HISA) 
              Author: Scott Longmore, Cooperative Institute for Research in the Atmosphere (CIRA) 
              Contact: Scott.Longmore@colostate.edu
              Last Edit: 2018-09-04


Version Number: v2.0.5

Release Date: beta 

Configuration Target:

  - Syntax

    /path/to/python2.7 /path/to/scripts/python/HISA/HISA.py -c /path/to/etc/HISA_<sat>.json -i /path/to/working/directory/HISA.PCF

  - Source Directory Structure

    src/
       makefile - Master makefile
       gfspack/ - gfspack code
       ShortTermTrack/ - ShortTermTrack code
       afdeck/ - afdeck code
       satcenter/ - satcenter code
       oparet/ - HISA algorithm code
       common/ - common source code 

    scripts/
       python/ - python scripts
          HISA/ - HISA system 
             HISA.py - HISA system driver
             libHISA.py - HISA library routines
             dataset.py - parent dataset class (abstract)
             ADECK.py - ATCF/ADECK dataset class 
             GFS.py - GFS dataset class
             ATMS.py - MiRS/ATMS dataset class
             NDE.py - NDE I/O library 
          pool/ - MIRS data subset extraction code
             data_pool.py - Extracts a subset of MIRS data from netCDF files to the query subdirectory
          convert/ - converts XYA/RZA files to netCDF
            convert2netCDF.py - netCDF converter
          plot/ - TC wind estimation plotting software
            main_read_plot_xya.py - Plots Wind Field Estimation File
          lib/ - library utility modules and routines 
          test/ - unit tests for system components
            fixtures/ - test data

    bin/ - compiled binaries
       bin2pack.x - converts GFS GRIB to CIRA pack format
       ShortTermTrack_mirs.x - find track from current or previous synoptic time
       afdeck.x - determines coordinates from track file
       satcenter.x -  determine satelite scanline time closest to storm
       HISA.x - hurricane intensity and strength estimation algorithm 
   
    etc/ - configuration and supplemenatry data files
          HISA_<sat>.json - HISA master configuration file (will be split into sub-files in future)
          mirs_atms_img.ini - configuration file for data pool
          oparet/oparet.cfg - scale, offset, and valid values for oparet input data
          oparet/coeffs/<instrument>
             oparet_params.cfg - main <instrument> config file
             pmn_atms0.inp.coef - Pressure miniumn coefficents?  
             r34_atms0.inp.coef - 34kt radius coefficients 
             r50_atms0.inp.coef - 50kt radius coefficients
             r64_atms0.inp.coef - 64kt radius coefficients
             vmx_atms0.inp.coef - Max velocity coeffcients
          convert/<sat>_xya2netCDF.json - XYA oparet text output to netCDF config
          convert/<sat>_rza2netCDF.json - RZA oparet text output to netCDF config
          plot/<sat>_plots_config.txt - ploting configuration file
    
    doc/ - documentation files
          StatusCodes.txt - general and script/executable specific status/exit/error codes

    test/ - component test directory

  - Working Directory Structure

    <RootDir>/ - Root Working Directory

       HISA.PCF - Process Control File, including all path, directory, external program, variables, settings, data filename, etc
       HISA.PSF - Process Status File
       HISA.LOG - Process Log File

       data/ - Input data files: storm track ATCF adeck (text), GFS (grib), and MIRS ATMS (netCDF) 

          <jtwc|nhc)_aBaSnYYYY.dat.YYYYmmddHHMMSS" - Input adeck file 
          <gfs.tHHz.pgrb2.1p00.fHHH.YYYYmmdd> - GFS analysis file (grib or pack)
          <NPR-MIRS-(SND|IMG)_vVVrR_Sat_sYYYYmmddHHMMFFF_eYYYYmmddHHMMFFF_cYYYYmmddHHMMFFF.nc> - MIRS data files (netCDF)

       model/ - Conversion of GFS grib to pack files
          *.bin - variable bin files
          AVN.DAT - GFS pack file
          
       database/ - MIRS subset data

       <BaSnYYYY>/ - Storm processing sub-directories
                     Ba - Basins: [al|ep|wp|cp|io|sh] 
                     Sn - Storm Number
                     YYYYY - Year

           ShortTermTrack_mirs.x - Find track from current or previous synoptic time
              <aBaSnYYYY.dat> - Input adeck file 
              <BaSnYYYY.inp>  - Output storm track file
             
           afdeck.x - Determines coordinates from track file
              <BaSnYYYY.inp> - Input storm track file
              COORDINATES - Output coordinates file

           satcenter.x - Determine satelite scanline time closest to storm
              <BaSnYYYY_YYYYMMDD>.<variable> - Input variable files from pool query (renamed with prefix)
              TIMES - Output time file

           HISA.x - Determines wind fields from MIRS satelite data and GFS model boundary conditions
              AVN.DAT - Input GFS model data file
              <variable.txt> - Input variable files linked to <BaSnYYYY_YYYYMMDD>.<variable> 
              COORTIMES - Input COORDINATE/TIMES concatenated file
              <BaSnYYYY_YYYYMMDD>.LOC - Output Location file 
              <BaSnYYYY_YYYYMMDD>.XYA - Output Wind Field Estimation file 
              <BaSnYYYY_YYYYMMDD>.STA - Output Statistics file 
              <BaSnYYYY_YYYYMMDD>.RZA - Output Variable file 
              <BaSnYYYY_YYYYMMDD>.FIX - Output Intensity/Size Estimation
              <BaSnYYYY_YYYYMMDD>.AFX - Output Aid File  
              <BaSnYYYY_YYYYMMDD>.log - Output Log file 

            main_read_plot_xya.py - Plots Wind Field Estimation File
              <BaSnYYYY_YYYYMMDD>.XYA - Input Wind Field Estimation file 
              <BaSnYYYY_YYYYMMDD>_T_ATMS_250mb.png -    Output 250mb temperature image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_200mb.png - Output 200mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_250mb.png - Output 250mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_300mb.png - Output 300mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_400mb.png - Output 400mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_500mb.png - Output 500mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_600mb.png - Output 600mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_700mb.png - Output 700mb wind field image file
              <BaSnYYYY_YYYYMMDD>_UV+Z_ATMS_800mb.png - Output 800mb wind field image file

       output/
         TC-<BaSnYYYY>_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.AFX - AMSU TC AFX fix file
         TC-<BaSnYYYY>_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS_<SnB>_FIX - AMSU TC (JTWC) fix file
         TC-<BaSnYYYY>-XYA_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.nc - XYA netCDF files 
         TC-<BaSnYYYY>-RZA_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.nc - RZA netCDF files 
         TC-<BaSnYYYY>-<field>-<level>_<ver>_<sat>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.png - Storm image files 

       log/
         <wgrib_YYmmddHH_F000_<fieldId/level>.log/err - wgrib sub-process log files
         <subprocess>_<stormId>.log/err - Sub-process storm loop log files
         <subprocess>_<stormId>_<satId>.log/err - Sub-process storm satellite log files

Reference Documents:


Compilation Instructions:

  The master makefile is located in the src/ directory. To compile simply run

     make

  To install executables in the bin/ directory

    make install

  - Compiler Warnings

    No compiler warning with gcc/gfortran 4.4.7-1

Product Control File Description:

    The following data files are needed for HISA: 

       -  The latest ATCDF adeck storm track text files avaiable from NHC/JTWC 
       -  The latest GFS/FV3 model analysis (F000) file in grib2 format 
       -  The latest set of MIRS (AMSU/ATMS) netCDF files within the last 9hr

Product Status File Description:
      
    The following products will be generated and staged in the working
    directory output/ subdirectory

       TC-<BaSnYYYY>_sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.<EXT> - Product files for each active storm defined in PSF
       TC-<BaSnYYYY>-<field>-<level>__sYYYYMMDDHHMMSSS_eYYYYMMDDHHMMSSS_cYYYYMMDDHHMMSSS.png - Product image for each active storm files defined in PSF

Production Rule Definitions:

    To be added


RELEASE_NOTES.TXT
RELEASE_NOTES.TXT -  CSU/CIRA/RAMMB Hurricane Intensity and Strength Algorithm (HISA) RELEASE_NOTES.TXT file
             
Author: Scott Longmore, Cooperative Institute for Research in the Atmosphere (CIRA) 
Contact: Scott.Longmore@colostate.edu
Last Edit: 2018-09-04

2018/09/04  v2.0.5 - Added HISA.json configs for SNPP/NOAA20, added/updated SNPP/NOAA20 netCDF convert json configs, 
                     added/updated SNPP/NOAA20 plot configs, added python requirements.txt/imports.txt, updated README.txt
2018/08/29  v2.0.4 - Converted oparet program to subroutine, added HISA driver, updated makefiles, HISA.json
2018/08/27  v2.0.3 - ATCF filter routine, GFS FV3 testing, dataset classes updates, naming changes, clean up 
2017/09/26  v2.0.2 - Fix plots coastline/boundries/meridians/parallels not being show, increased map resolution to high
2017/09/25  v2.0.1 - Fix for query file gather for correct ordered files, scaling/offset/missing values corrections 
2017/07/10  v2.0.0 - New python driver (mirs_tc.py), new data pool (scripts/python/pool), new operat (src/operat) 
2015/05/15  v1.5.4 - Removed legacy version of nbevs routine insrc/oparet/nbevs.f90, updated besubs.f90 
2015/05/15  v1.5.3 - OPERATIONAL RELEASE
                   - Logging update requested by NDE, WARNING issued when pool_query.py finds no matching records,d
                     sub-programs that exit normally but do not produce output produce WARNING. 
                   - QC.TXT link issues for multiple storms fixed
                   - oparet coefficients updated, minor updates
                   - JPSS satellite variables (for future) commented out
2015/02/02  v1.5.2 - GFS file naming update 
2014/12/10  v1.5.1 - oparet fix for SH storms
                     ShortTermTrack fix for invest storms (missing positions)
                     netCDF global attribute fixes/updates 
2014/11/09  v1.5.0 - Added ATMS QC > 1 for +20% of retrievals, exits algorithm 
                   - Fix for M12 of ShortTermTrack to current synoptic time
                   - ATMS time windows adjusted, gettimes.sh replaced with datetimes.py
		   - XYA/RZA netCDF CF compliance fixes, standard names, valid range, units
		   - JTWC ATCF/FIX filename convention added to products
                   - npp_tc.py scheduler added, eventually will replace master script NPP_TC.sh
2014/07/30  v1.4.2 - Problems with new graphics code on newer ATMS MIRS data 
                     reverted to previous version of plotting code
2014/07/20  v1.4.1 - Added non-zero exit code status in ShortTermTrack for corrupt ATCF files 
                     Updated plotting code, graphics fill more of the frame
2014/07/03  v1.4.0 - Algorithm replaced MIRS ATMS cloud liquid water (clw) with liquid water path (lwp) 
                     to incorporate precipitating (eyewall/rainband) retrievals and added bias correction.
                   - Some sub-program (ShortTermTrack, Oparet) log designators changed from WARNING to INFO 
                     for non-critical exceptions (00 or -12 ATCF fix missing, swath too far from storm, etc) 
2014/05/01  v1.3.4 - Removed dependency on MIRS ATMS attributes with data_pool and oparat changes
                     Added etc/oparet.cfg for correct scaling and offsets
                     Removed TC Sandy case setup/clean scripts, TC test case scripts included with 
                     seperate case data tar file
2014/03/06  v1.3.3 - PCF validator, removes older adeck/gfs files, fix for leap year in NPP_TC_GETTIMES.sh
2014/02/08  v1.3.2 - netCDF bug fixes, minor updates 
                   - ERROR to WARNING messages for sub-programs inside storm loop, so if one TC fails,
                     remaining TC's are processed 
2014/02/02  v1.3.1 - netCDF converter for RZA files
2014/01/27  v1.3.0 - netCDF converter for XYA files, MiRS file fix
2014/01/07: v1.2.3 - ATCF (adeck), GFS grib file naming, product version in file names
2013/11/14: v1.2.2 - Data test case inclusion
2013/11/08: v1.2.1 - PCF formatting, required NDE and data input variable upgrades 
                   - Data, temporary sub-directory hierarchy handling
                   - Static variables to master script, external and data variables to PCF
                   - Date format/handling between NDE and master script
                   - Executable/script location execution handling from working to operation directory 
                   - PSF output file location and naming is now absolute 
                   - LOG file location to top working directory
2013/10/31: v1.2.0 - NDE framework upgrades: PCF, PSF, logging, exception handling
                   - GFS grib to pack file converter
2013/06/25: v1.1.0 - NDE initial version
2013/06/14: v1.0.0 - Initial version


scripts/python/HISA/ATMS.py
#!/usr/bin/python
"""
ATMS.py - MIRS SNPP ATMS class 
"""

# Stock modules
import sys
import os
import re
import logging
import traceback
import datetime
import collections
import operator

# Local modules
import error_codes
import utils
from dataset import dataset

LOG = logging.getLogger(__name__)

# NPR-MIRS-SND_v9_NPP_s201208180059460_e201208180100176_c20130308234921.nc
# NPR-MIRS-IMG_v9_NPP_s201208180058100_e201208180058416_c20130308234421.nc
fileFieldsSchema={
    "$schema": "http://json-schema.org/draft-04/schema#",
    "type":"object",
    "properties": {
        "type":{ "type":"string" },
        "version":{ "type":"string" },
        "satellite":{ "type":"string" },
        "startDTG":{ "type":"string" },
        "endDTG":{ "type":"string" },
        "createDTG":{ "type":"string" }
    }
}
datasets={
    "mirs_atms_img":{
       "type":"IMG",
    },
    "mirs_atms_snd":{
       "type":"SND",
    }
}
dtFormat="%Y%m%d%H%M%S"

class atms(dataset):

    def __init__(self,properties):

        self.schema=fileFieldsSchema

        super(atms,self).__init__(properties)

        # Seconds or milliseconds
        for dt in ['startDT','endDT','createDT']:
            if len(self._properties[dt]) == 15:
                self._properties[dt]=datetime.datetime.strptime(self._properties[dt][:-1],dtFormat)
            else:
                self._properties[dt]=datetime.datetime.strptime(self._properties[dt],dtFormat)

        return


def filterAtms(config,metadata,dataname):
   
    filterAtms=[]
    dataset=config['datasets'][dataname]
    if dataset['datalink'] in config['datalinks']:

        datalink=config['datalinks'][dataset['datalink']]
        datasetKeys=datalink['datasetKeys']
        links=datalink['links']

        try:
            for atms in metadata[dataname]:
                linkID=atms.get(datasetKeys[dataname])
                if linkID in links: 
                    filterAtms.append(atms)
                else:
                    utils.moveFile(atms.get('path'),atms.get('filename'),config['dirs']['discard'])
                    LOG.info("Moving unmatched file: {} to discard dir: {}".format(atms.get('filename'),config['dirs']['discard']))
        except:
            msg="Error matching/filtering dataset: {}".format(dataset)
            utils.error(LOG,msg,error_codes.EX_IOERR)

    else:
       msg="config[\"datalinks\"][{}] definition required, see MIRS_TC.datalinks".format(dataset['datalink'])
       utils.error(LOG,msg,error_codes.EX_IOERR)

    return(filterAtms)
